opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,154 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=abstract-class-instantiated
+# pylint: disable=too-many-ancestors
+# pylint: disable=useless-super-delegation
+
+# type: ignore
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from re import ASCII
+from re import compile as compile_
+from typing import Callable
+
+_logger = getLogger(__name__)
+
+
+class Instrument(ABC):
+
+    _name_regex = compile_(r"[a-zA-Z][-.\w]{0,62}", ASCII)
+
+    @abstractmethod
+    def __init__(self, name, *args, unit="", description="", **kwargs):
+
+        if self._name_regex.fullmatch(name) is None:
+            raise Exception("Invalid instrument name {}".format(name))
+
+        if len(unit) > 63:
+            raise Exception("unit must be 63 characters or shorter")
+
+        if any(ord(character) > 127 for character in unit):
```

Comment:
I'm not too familiar with `re.compile` with `ASCII` but does that not already take care of this check?

Reply:
Hmno, it does not:

```  python
n [16]: from re import ASCII, compile

In [17]: _name_regex = compile(r"[a-zA-Z][-.\w]{0,62}", ASCII)

In [18]: _name_regex.match("aaa")
Out[18]: <re.Match object; span=(0, 3), match='aaa'>

In [19]: _name_regex.match("aÑa")
Out[19]: <re.Match object; span=(0, 1), match='a'>

In [20]: ord("a")
Out[20]: 97

In [21]: ord("Ñ")
Out[21]: 209
```
I'll remove the `ASCII` flag.